### DISCUSSION TITLE
How to hide an entry from the Context Menu

### DISCUSSION THREAD

**Discussion by lgmarin** [TypeFox/monaco-languageclient#580] [Q&A]
Hi!

I'm having some troubles since I updated from the older versions of the Language Client, before, I was able to remove some of the Context Menu items that I will not support on my Language Server (I'm using python-lsp-server with a custom plugin).

Since the update I was able to go with this [solution](https://github.com/microsoft/monaco-editor/issues/1567#issuecomment-1404897422) but now it does not work.

I tried to find other ways through the API, as the examples use vscode.commands but it does not seem that they have access to the same data.

One thing that I tried was to override the command Id that I don't want to show with contextMenuGroupId set to undefined, but it also didn't work, actually it seems that it will not override the command at all.

```typescript
    public removeMenuItems(removableIds: Array<string>) {
        for (let removeId in removableIds)
        { 
            this.getEditor()?.addAction({
                id: removeId, label: '_', run: ()=>{},
                keybindings: [],
                contextMenuOrder: 0, contextMenuGroupId: undefined,
            });
        }
    }
```

**Comment by CGNonofr** [comment]
Out of curiosity, what menu items do you wish to remove?

**Reply by lgmarin** [reply]
I'm trying to remove those:

```typescript
            "editor.action.refactor",
            "editor.action.revealDefinition",
            "editor.action.goToTypeDefinition",
            "editor.action.goToImplementation",
            "editor.action.goToReferences",
            "editor.action.revealDeclaration",
            "editor.action.formatDocument",
            "editor.action.formatSelection",
            "editor.action.rename",
            "Peek",
            "EditorContextPeek"
```

**Reply by CGNonofr** [reply]
Most of those menus are not supposed to appear if there is no provider registered, so why are they visible in your case?

**Reply by lgmarin** [reply]
They appear when the Language Server is connected with Websocket. 

I may have another question then, I tried to find a way to disable some of the capabilities from the Client side, is it possible? I tried to use handleRegisterCapability middleware but nothing happens there. Is there a way to disable or select the capabilities to be used by the editor <> LSP?

I guess this would make more sense.

**Reply by CGNonofr** [reply]
What reason would make you want to disable features supported by both server and client? This library used to be able to disable unsupported feature but I don't think it's possible anymore, right @kaisalmen ?

**Reply by lgmarin** [reply]
An example is the goToImplementation, it is not working, as I don't have access to the file system, but right now, it is not really a issue for me and I don't need this feature. That's why before I was just removing the items from the menu, for my purposes, that was enough.

And other thing, for the Format Document I had a callback that would handle that for me, so I removed from the menu, and then added another action for Format Document, but with a different run: action.

**Reply by CGNonofr** [reply]
Maybe the best solution is to use a language client middleware then: intercept the server capabilities and remove those you don't want

**Reply by lgmarin** [reply]
I tried this, with no luck so far

```typescript
                middleware: {
                    didOpen: async (data: any, next: any) => {
                        next(data);
                        this.updateState('open');
                        this._logger?.info("Received document/didOpen from the Language Server.");
                    },
                    handleRegisterCapability: async (params: any, next: any) => {
                        console.log('Capabilities: ');
                        console.log(params);
                        next(params);
                    },
                }
``` 

The didOpen works just fine, but the handleRegisterCapability is never called.

Thank you for @CGNonofr for the help. I know that this kind of a weird use case, but is the only thing holding me since I updated the packages.

**Reply by CGNonofr** [reply]
I think `handleRegisterCapability` is only called for dynamic capabilities

I don't think you'll be able to do it using the language client middleware directly, you'll probably need to intercept the `initialize` response at the protocol level to change the capabilities

**Reply by kaisalmen** [reply]
> This library used to be able to disable unsupported feature but I don't think it's possible anymore, right @kaisalmen ?

Yeah, this code is gone. Everything `vscode-languageclient` provides is available. You already outlined possible ways forward. Another possibility is to bring back that old code (not in this lib, but in user/app code). It used to look like this: https://github.com/TypeFox/monaco-languageclient/blob/mlc-6.2.0/packages/client/src/monaco-language-client.ts

**Reply by lgmarin** [reply]
Ok, thank you for the help!

Just one more thing, is it possible to create a custom ContextMenu? Anyways, would be really nice if the ContextMenu where more configurable.

**Reply by CGNonofr** [reply]
You can add additional context menus using the VSCode api via a VSCode extension (see https://code.visualstudio.com/api/references/contribution-points#contributes.menus)
